### PR TITLE
feat: update firecrawl tool with additional features

### LINKED ISSUE #1349: Update Firecrawl tool with `Async Crawling` and `Map` functionality.
current implementation of firecrawl tool within `phidata` includes `scrape_url` and `crawl_url` methods.
However, there are few other new methods / features introduced in the original python SDK like `async crawling` and `map` method.
i would like to include them in the phidata firecrawl tool since I believe these updates will be valuable to users.

To understand more about the features, one can refer this:
https://docs.firecrawl.dev/sdks/python

pls lemme know how we take this forward or if there's any clarification required at my end.

ccing: @manthanguptaa  for speeding up the process.

### PR DESCRIPTION
**Pull request by AbhishekRP2002**
- fresh clean PR aiming to close #1349 
- Previous closed PR for ref: #1357 

cc: @manthanguptaa

### CODE REVIEW

**Comment by AbhishekRP2002** [comment]
@manthanguptaa  any updates on this ?

**Comment by manthanguptaa** [comment]
Ran into this issue while running the code

Screenshot 2024-11-19 at 2 38 18 PM

File: cookbook/tools/firecrawl_tools.py [hunk 1]
```diff
@@ -1,5 +1,31 @@
...
+
+
+def crawl_example(url: str):
+    agent = Agent(
+        tools=[FirecrawlTools(scrape=True, crawl=True, async_crawl=True, limit=25)],
+        show_tool_calls=True,
+        reasoning=True,
+        markdown=True,
+    )
+    agent.print_response(f"Crawl and analyze the main use cases and  insights using asynchronous crawling: {url}")
```
Comment (manthanguptaa) [inline cookbook/tools/firecrawl_tools.py:25]: extra whitespace in the middle of the sentence

File: cookbook/tools/firecrawl_tools.py
Comment (manthanguptaa) [inline cookbook/tools/firecrawl_tools.py:1]: Can you not have Agents in functions and then call them? I think you can have only 1 Agent with the following config
```
Agent(
        tools=[FirecrawlTools(scrape=True, crawl=True, async_crawl=True, limit=25)],
        show_tool_calls=True,
        reasoning=True,
        markdown=True,
    )
```
and then ask the agent to do it synchronously or asynchronously

File: phi/tools/firecrawl.py [hunk 2]
```diff
@@ -9,33 +13,47 @@
...
-            crawl = True
-
-        self.register(self.scrape_website)
-        self.register(self.crawl_website)
+            self.register(self.crawl_website)
+        if async_crawl:
+            self.register(self.async_crawl_website)
+        if crawl or async_crawl:
+            self.register(self.check_crawl_status)
+            self.register(self.cancel_crawl)
```
Comment (manthanguptaa) [inline phi/tools/firecrawl.py:50]: async crawl cannot be cancelled?

File: phi/tools/firecrawl.py [hunk 3]
```diff
@@ -9,33 +13,47 @@
...
-        elif not scrape:
-            crawl = True
-
-        self.register(self.scrape_website)
-        self.register(self.crawl_website)
+            self.register(self.crawl_website)
+        if async_crawl:
+            self.register(self.async_crawl_website)
+        if crawl or async_crawl:
+            self.register(self.check_crawl_status)
```
[the remark this answers is not in the document]
Comment (AbhishekRP2002) [reply phi/tools/firecrawl.py:49]: if only `async_crawl` is True then also the code block will execute @manthanguptaa

**Comment by AbhishekRP2002** [comment]
> Ran into this issue while running the code
> 
> Screenshot 2024-11-19 at 2 38 18 PM

@manthanguptaa  did u encounter it while running the cookbook script ?
what is the openai model being used by default ?
This is most likely due to stochastic nature of the llm model, as a result of which it could parse the exact tool params . I wasn't able to reproduce it. 

PFA for ref:
image
image
image
image

**Comment by AbhishekRP2002** [comment]
Hi @manthanguptaa any updates on this ? Please lemme know if there's refactoring that needs to be done.

**Review by manthanguptaa — CHANGES_REQUESTED** [review]

**Comment by manthanguptaa** [comment]
Took this to completion in #1659

**Pull request closed without merging**